controlplane/pkg/nsmd/nsmd.go

```diff
@@ -170,120 +170,180 @@ func (nsm *nsmServer) EnumConnection(context context.Context, request *nsmdapi.E
-						}
-					}
-
-					if err != nil {
-						logrus.Errorf("Failed to register NSE: %v", err)
-					} else {
-						updatedNSEs[newReg.NetworkserviceEndpoint.EndpointName] = nseregistry.NSEEntry{
-							Workspace: ws.Name(),
-							NseReg:    newReg,
-						}
-					}
-				} else {
-					nse.NseReg.NetworkServiceManager = nsm.model.GetNsm()
-					nse.NseReg.NetworkserviceEndpoint.NetworkServiceManagerName = nse.NseReg.NetworkServiceManager.Name
-					nsm.model.AddEndpoint(&model.Endpoint{
-						Endpoint:       nse.NseReg,
-						Workspace:      nse.Workspace,
-						SocketLocation: ws.NsmClientSocket(),
-					})
-					updatedNSEs[endpointId] = nse
-				}
+	networkServices := map[string]bool{}
+	updatedNSEs := map[string]nseregistry.NSEEntry{}
+	for name, nse := range nses {
+		ws, ok := nsm.workspaces[nse.Workspace]
+		if !ok {
+			continue
+		}
+
+		logrus.Infof("Checking NSE %s is alive at %v...", name, ws.NsmClientSocket())
+		if !ws.isConnectionAlive(NSEAliveTimeout) {
+			logrus.Errorf("Unable to connect to local nse %v. Skipping", nse.NseReg)
+			if err := nsm.deleteEndpointWithClient(name, registryClient); err != nil {
+				logrus.Errorf("Remove NSE: NSE %v", err)
 			}
+			continue
 		}
-	} else {
-		// We do not have NSE's, need to unregister all of them.
-		// Restore NSEs
-		client, err := nsm.serviceRegistry.NseRegistryClient()
+
+		logrus.Infof("NSE %s is alive at %v...", name, ws.NsmClientSocket())
+		newName, newNSE, err := nsm.restoreEndpoint(discoveryClient, registryClient, name, nse, ws, registeredNSEs, networkServices)
 		if err != nil {
-			err = fmt.Errorf("Failed to get RegistryClient: %s", err)
-			return
+			logrus.Errorf("Failed to register NSE: %v", err)
+			continue
 		}
 
-		for _, nse := range existingEndpoints {
-			if _, err := client.RemoveNSE(context.Background(), &registry.RemoveNSERequest{
-				EndpointName: nse.EndpointName,
+		networkServices[newNSE.NseReg.GetNetworkService().GetName()] = true
+		updatedNSEs[newName] = newNSE
+	}
+
+	// We need to unregister NSE's without NSM registration.
+	for name := range registeredNSEs {
+		if _, ok := updatedNSEs[name]; !ok {
+			if _, err := registryClient.RemoveNSE(context.Background(), &registry.RemoveNSERequest{
+				EndpointName: name,
 			}); err != nil {
 				logrus.Errorf("Remove NSE: NSE %v", err)
 			}
 		}
 	}
-	if len(updatedClients) > 0 || len(updatedNSEs) > 0 {
-		if err := nsm.localRegistry.Save(updatedClients, updatedNSEs); err != nil {
-			logrus.Errorf("Store updated NSE local registry...")
+
+	return updatedNSEs, nil
+}
+
+func (nsm *nsmServer) restoreEndpoint(
+	discoveryClient registry.NetworkServiceDiscoveryClient,
+	registryClient registry.NetworkServiceRegistryClient,
+	name string,
+	nse nseregistry.NSEEntry,
+	ws *Workspace,
+	registeredNSEs map[string]string,
+	networkServices map[string]bool) (string, nseregistry.NSEEntry, error) {
+
+	if networkService, ok := registeredNSEs[name]; ok {
+		if networkServices[networkService] {
```

Comment:
If we have 2 NSE that implement the same NS, and the second NSE is absent in registry. Does it mean that we will not register the second NSE because of this if condition?

Reply:
`registeredNSEs` are `NSE` resources from registry. We only need to reregister such NSE if there is no corresponding `networkservice` resource in registry.